src/tasks/getsubscriptiondetails/taskgetsubscriptiondetails.cpp

```diff
@@ -0,0 +1,117 @@
+
+#include "taskgetsubscriptiondetails.h"
+#include "authenticationlistener.h"
+#include "authenticationinapp/authenticationinappsession.h"
+#include "errorhandler.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+namespace {
+Logger logger(LOG_MAIN, "TaskGetSubscriptionDetails");
+}  // anonymous namespace
+
+TaskGetSubscriptionDetails::TaskGetSubscriptionDetails(
+    const QString& emailAddress)
+    : Task("TaskGetSubscriptionDetails"), m_emailAddress(emailAddress) {
+  MVPN_COUNT_CTOR(TaskGetSubscriptionDetails);
+}
+
+TaskGetSubscriptionDetails::~TaskGetSubscriptionDetails() {
+  MVPN_COUNT_DTOR(TaskGetSubscriptionDetails);
+}
+
+void TaskGetSubscriptionDetails::run() {
+  logger.debug() << "run";
+
+  NetworkRequest* request =
+      NetworkRequest::createForGetSubscriptionDetails(this);
+
+  connect(
+      request, &NetworkRequest::requestFailed, this,
+      [this](QNetworkReply::NetworkError error, const QByteArray&) {
+        logger.error() << "Get subscription details failed" << error;
+
+        // Network request failed after authentication for a second time
+        if (m_authenticationInAppSession) {
+          MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+          m_authenticationInAppSession->terminate();
+          return;
+        }
+
+        // User needs to (re)authenticate
+        if (error == QNetworkReply::AuthenticationRequiredError) {
+          needsAuthentication();
+          return;
+        }
+
+        MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+        emit completed();
+      });
+
+  connect(request, &NetworkRequest::requestCompleted, this,
+          [this](const QByteArray& data) {
+            logger.debug() << "Get subscription details completed" << data;
+
+            MozillaVPN::instance()->subscriptionDetailsFetched(data);
+
+            if (m_authenticationInAppSession) {
+              m_authenticationInAppSession->terminate();
+              return;
+            }
+            emit completed();
+          });
+}
+
+void TaskGetSubscriptionDetails::needsAuthentication() {
+  Q_ASSERT(!m_authenticationInAppSession);
+
+  QByteArray pkceCodeVerifier;
+  QByteArray pkceCodeChallenge;
+  AuthenticationListener::generatePkceCodes(pkceCodeVerifier,
+                                            pkceCodeChallenge);
+  Q_ASSERT(!pkceCodeVerifier.isEmpty() && !pkceCodeChallenge.isEmpty());
+
+  m_authenticationInAppSession = new AuthenticationInAppSession(this);
+
+  connect(m_authenticationInAppSession, &AuthenticationInAppSession::terminated,
+          this, &Task::completed);
+
+  connect(m_authenticationInAppSession, &AuthenticationInAppSession::completed,
+          this, [this, pkceCodeVerifier](const QString& pkceCodeSuccess) {
+            logger.debug() << "Authentication completed with code:"
+                           << pkceCodeSuccess;
```

Comment:
logger.`debug.sensitive()` here and in the authenticate.cpp - we don't need to write it in prod (it's an existing bug, but let's fixing in this PR)